include/pybind11/pytypes.h

```diff
@@ -1637,7 +1639,39 @@ class capsule : public object {
         }
     }
 
-    const char *name() const { return PyCapsule_GetName(m_ptr); }
+    const char *name() const {
+        const char *name = PyCapsule_GetName(m_ptr);
+        if ((name == nullptr) && PyErr_Occurred()) {
+            throw error_already_set();
+        }
+        return name;
+    }
+
+    /// Replaces a capsule's name *without* calling the destructor on the existing one.
+    void set_name(const char *new_name) {
+        if (PyCapsule_SetName(m_ptr, new_name) != 0) {
+            throw error_already_set();
+        }
+    }
+
+private:
+    static const char *get_name_or_throw(PyObject *o) {
+        /* an exception may be in-flight, we must save it in case we create another one */
+        PyObject *type = nullptr, *value = nullptr, *traceback = nullptr;
+        PyErr_Fetch(&type, &value, &traceback);
```

Comment:
We have `error_scope` in detail/common.h. Do you think that could work here? My guess is it does.
(Although now that I look at it, we're currently not following the rule-of-3 in the `error_scope` implementation, but that's best fixed in a separate PR IMO.)

Reply:
Very nice! Thanks for the suggestion @rwkg. It totally belongs here and make code much cleaner too.